{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "import re\n",
    "import itertools\n",
    "import numpy\n",
    "\n",
    "def loadFasta(filename):\n",
    "    if (filename.endswith(\".gz\")):\n",
    "        fp = gzip.open(filename, 'rb')\n",
    "    else:\n",
    "        fp = open(filename, 'rb')\n",
    "    # split at headers\n",
    "    data = fp.read().split('>')\n",
    "    fp.close()\n",
    "    # ignore whatever appears before the 1st header\n",
    "    data.pop(0)     \n",
    "    headers = []\n",
    "    sequences = []\n",
    "    for sequence in data:\n",
    "        lines = sequence.split('\\n')\n",
    "        headers.append(lines.pop(0))\n",
    "        # add an extra \"+\" to make string \"1-referenced\"\n",
    "        sequences.append('+' + ''.join(lines))\n",
    "    return (headers, sequences)\n",
    "\n",
    "def createBinaryVectorFile(kmerList):\n",
    "    baseMatrices = {'A': numpy.asarray([0,0,0,1]).T, 'C': numpy.asarray([0,0,1,0]).T, 'G': numpy.asarray([0,1,0,0]).T, 'T': numpy.asarray([1,0,0,0]).T}\n",
    "    binFile = open('kmerBinary.csv', 'w')\n",
    "    for key, value in kmerList.iteritems():\n",
    "        for kmer in value:\n",
    "            for character in kmer:\n",
    "                binFile.write(baseMatrices.get(character))\n",
    "                binFile.write(\",\")\n",
    "            binFile.write(key)\n",
    "            binFile.write(\"\\n\")\n",
    "    binFile.close()\n",
    "\n",
    "def createHistogramFile(kmerList, histKmerLen):\n",
    "    histFile = open('kmerHistograms.csv', 'w')\n",
    "    motifs = {}\n",
    "    motifList = []\n",
    "    \n",
    "    for pattern in itertools.product('ACTG', repeat=histKmerLen):\n",
    "        motif = ''.join(pattern)\n",
    "        motifs[motif] = 0\n",
    "        motifList.append(motif)\n",
    "        histFile.write(motif)\n",
    "        histFile.write(\",\")\n",
    "    \n",
    "    histFile.write(\"\\n\")\n",
    "    freqKmers = {}\n",
    "    \n",
    "    for key, value in kmerList.iteritems():\n",
    "        for kmer in value:\n",
    "            for i in xrange(len(kmer)-histKmerLen+1):\n",
    "                histKmer = kmer[i:i+histKmerLen]\n",
    "#                 motifFreq =  motifs.get(histKmer, 0)\n",
    "#                 motifFreq += 1\n",
    "                motifs[histKmer] = motifs.get(histKmer, 0) + 1\n",
    "        maxFreq = 0\n",
    "        maxKmer = ''\n",
    "\n",
    "        for motif in motifList:\n",
    "            if motifs[motif] > maxFreq:\n",
    "                maxFreq = motifs[motif]\n",
    "                maxKmer = motif\n",
    "            histFile.write(str(motifs[motif]))\n",
    "            histFile.write(\",\")\n",
    "            motifs[motif] = 0\n",
    "        freqKmers[maxKmer] = freqKmers.get(maxKmer, 0)+1\n",
    "            \n",
    "        histFile.write(key)\n",
    "        histFile.write(\"\\n\")\n",
    "\n",
    "#    print freqKmers\n",
    "    histFile.close()\n",
    "    \n",
    "def callFasta(fileName, baseCoverage, longKmerLen, shortKmerLen, histKmerLen):\n",
    "    header, seq = loadFasta(fileName)\n",
    "\n",
    "    transcriptLocation = 1\n",
    "    kmerList = {}\n",
    "    for i in xrange(len(header)):\n",
    "        transcriptID = re.split('\\|+', header[i])[transcriptLocation][7:]\n",
    "        kmerListForTranscript = []\n",
    "#         for j in xrange(1, len(seq[i])-longKmerLen+1, (longKmerLen/baseCoverage)):\n",
    "        for j in xrange(1, len(seq[i])-longKmerLen+1):\n",
    "            kmer = seq[i][j:j+longKmerLen]\n",
    "            for k in xrange(len(kmer)-shortKmerLen+1):\n",
    "                shortKmer = kmer[k:k+shortKmerLen]\n",
    "                kmerListForTranscript.append(shortKmer)\n",
    "\n",
    "#         if (len(seq[i])-1) > longKmerLen and (len(seq[i])-1)%(longKmerLen/baseCoverage) != 0:\n",
    "#             kmer = seq[i][-longKmerLen:]\n",
    "#             for k in xrange(len(kmer)-shortKmerLen+1):\n",
    "#                 shortKmer = kmer[k:k+shortKmerLen]\n",
    "#                 kmerListForTranscript.append(shortKmer)\n",
    "        if len(seq[i])-1 <= longKmerLen:\n",
    "            kmer = seq[i]\n",
    "            for k in xrange(len(kmer)-shortKmerLen+1):\n",
    "                shortKmer = kmer[k:k+shortKmerLen]\n",
    "                kmerListForTranscript.append(shortKmer)\n",
    "\n",
    "        kmerList[transcriptID] = kmerListForTranscript\n",
    "\n",
    "    createBinaryVectorFile(kmerList)\n",
    "#    createHistogramFile(kmerList, histKmerLen)\n",
    "\n",
    "fileName = \"chr1_0_10mil_GID_TID_cDNA_strand.txt\" \n",
    "baseCoverage = 5\n",
    "longKmerLen = 100\n",
    "shortKmerLen = 31\n",
    "histKmerLen = 5\n",
    "tlengs = callFasta(fileName, baseCoverage, longKmerLen, shortKmerLen, histKmerLen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.394447004647\n",
      "0.728162994921\n",
      "0.84468014014\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import itertools\n",
    "import numpy\n",
    "import random\n",
    "\n",
    "def loadFasta(filename):\n",
    "    if (filename.endswith(\".gz\")):\n",
    "        fp = gzip.open(filename, 'rb')\n",
    "    else:\n",
    "        fp = open(filename, 'rb')\n",
    "    # split at headers\n",
    "    data = fp.read().split('>')\n",
    "    fp.close()\n",
    "    # ignore whatever appears before the 1st header\n",
    "    data.pop(0)     \n",
    "    headers = []\n",
    "    sequences = []\n",
    "    for sequence in data:\n",
    "        lines = sequence.split('\\n')\n",
    "        headers.append(lines.pop(0))\n",
    "        # add an extra \"+\" to make string \"1-referenced\"\n",
    "        sequences.append('+' + ''.join(lines))\n",
    "    return (headers, sequences)\n",
    "\n",
    "def createHistogramFile(kmerList, histKmerLen):\n",
    "    histFile = open('kmerHistograms.csv', 'w')\n",
    "    motifs = {}\n",
    "    motifList = []\n",
    "    \n",
    "    for pattern in itertools.product('ACTG', repeat=histKmerLen):\n",
    "        motif = ''.join(pattern)\n",
    "        motifs[motif] = 0\n",
    "        motifList.append(motif)\n",
    "        histFile.write(motif)\n",
    "        histFile.write(\",\")\n",
    "    \n",
    "    histFile.write(\"\\n\")\n",
    "    freqKmers = {}\n",
    "    \n",
    "    for key, value in kmerList.iteritems():\n",
    "        for kmer in value:\n",
    "            for i in xrange(len(kmer)-histKmerLen+1):\n",
    "                histKmer = kmer[i:i+histKmerLen]\n",
    "#                 motifFreq =  motifs.get(histKmer, 0)\n",
    "#                 motifFreq += 1\n",
    "                motifs[histKmer] = motifs.get(histKmer, 0) + 1\n",
    "        maxFreq = 0\n",
    "        maxKmer = ''\n",
    "\n",
    "        for motif in motifList:\n",
    "            if motifs[motif] > maxFreq:\n",
    "                maxFreq = motifs[motif]\n",
    "                maxKmer = motif\n",
    "            histFile.write(str(motifs[motif]))\n",
    "            histFile.write(\",\")\n",
    "            motifs[motif] = 0\n",
    "        freqKmers[maxKmer] = freqKmers.get(maxKmer, 0)+1\n",
    "            \n",
    "        histFile.write(key)\n",
    "        histFile.write(\"\\n\")\n",
    "\n",
    "#    print freqKmers\n",
    "    histFile.close()\n",
    "\n",
    "def createBinaryVector(kmer):\n",
    "    baseMatrices = {'A': '0001', 'C': '0010', 'G': '0100', 'T': '1000'}\n",
    "    newKmer = ''\n",
    "    for character in kmer:\n",
    "        newKmer += baseMatrices.get(character)\n",
    "    return newKmer\n",
    "\n",
    "def callFasta(fileName, baseCoverage, longKmerLen, shortKmerLen, histKmerLen):\n",
    "    header, seq = loadFasta(fileName)\n",
    "\n",
    "    transcriptLocation = 1\n",
    "    kmers = []\n",
    "    transcripts = []\n",
    "    validation_kmers = []\n",
    "    validation_transcripts = []\n",
    "    test_kmers = []\n",
    "    test_transcripts = []\n",
    "    \n",
    "    for i in xrange(len(header)):\n",
    "        transcriptID = re.split('\\|+', header[i])[transcriptLocation][7:]\n",
    "        for j in xrange(1, len(seq[i])-longKmerLen+1):\n",
    "            kmer = seq[i][j:j+longKmerLen]\n",
    "            for k in xrange(len(kmer)-shortKmerLen+1):\n",
    "                shortKmer = createBinaryVector(kmer[k:k+shortKmerLen])\n",
    "                kmers.append(shortKmer)\n",
    "                transcripts.append(transcriptID)\n",
    "                val = random.random()\n",
    "                if val <= 0.20:\n",
    "                    validation_kmers.append(shortKmer)\n",
    "                    validation_transcripts.append(transcriptID)\n",
    "                if val >= .80:\n",
    "                    test_kmers.append(shortKmer)\n",
    "                    test_transcripts.append(transcriptID)\n",
    "\n",
    "        if len(seq[i])-1 <= longKmerLen:\n",
    "            kmer = seq[i]\n",
    "            for k in xrange(len(kmer)-shortKmerLen+1):\n",
    "                shortKmer = createBinaryVector(kmer[k:k+shortKmerLen])\n",
    "                kmers.append(shortKmer)\n",
    "                transcripts.append(transcriptID)\n",
    "                val = random.random()\n",
    "                if val <= 0.20:\n",
    "                    validation_kmers.append(shortKmer)\n",
    "                    validation_transcripts.append(transcriptID)\n",
    "                if val >= .80:\n",
    "                    test_kmers.append(shortKmer)\n",
    "                    test_transcripts.append(transcriptID)\n",
    "\n",
    "    return kmers, transcripts, validation_kmers, validation_transcripts, test_kmers, test_transcripts\n",
    "\n",
    "#    createHistogramFile(kmerList, histKmerLen)\n",
    "\n",
    "fileName = \"chr1_0_10mil_GID_TID_cDNA_strand_short.txt\" \n",
    "baseCoverage = 5\n",
    "longKmerLen = 100\n",
    "shortKmerLen = 31\n",
    "histKmerLen = 5\n",
    "kmers, transcripts, validation_kmers, validation_transcripts, test_kmers, test_transcripts = callFasta(fileName, baseCoverage, longKmerLen, shortKmerLen, histKmerLen)\n",
    "print len(kmers)\n",
    "print len(transcripts)\n",
    "print len(validation_kmers)\n",
    "print len(validation_transcripts)\n",
    "print len(test_kmers)\n",
    "print len(test_transcripts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "109\n"
     ]
    }
   ],
   "source": [
    "temp = {'AAATG': 2, 'GCCCA': 2, 'GCCCC': 1, 'AAATA': 1, 'AAATT': 1, 'TGTCT': 2, 'CAAAA': 1, 'CAAAG': 1, 'AGTAC': 1, 'CCGCC': 3, 'CCTAG': 1, 'TGGTG': 3, 'CGCAA': 1, 'GTGGT': 1, 'TCTGC': 2, 'GCTGT': 1, 'TCTGG': 2, 'CCCCA': 1, 'CCCCC': 1, 'AAGAA': 10, 'AAGAG': 1, 'AGCAG': 2, 'ACAGA': 1, 'CAGCA': 1, 'CTGGA': 2, 'CTGGC': 1, 'TCATC': 1, 'GAAAA': 6, 'TATAT': 1, 'AACAA': 2, 'TGAAG': 2, 'GAGTT': 1, 'CCTGG': 1, 'TGATG': 1, 'CCTTT': 1, 'GGCCG': 1, 'CCTTC': 1, 'CAAGA': 3, 'AGAAT': 1, 'AGAAG': 2, 'AGAAA': 9, 'TTAAA': 2, 'GCTTC': 1, 'AAGGA': 1, 'ATGCC': 1, 'AAGGC': 1, 'AATGG': 1, 'TTCCA': 1, 'GAAGA': 11, 'AATGA': 2, 'AGCCT': 2, 'CTCTG': 1, 'CCACA': 1, 'AATTC': 1, 'CTCCC': 1, 'GAGGA': 1, 'AAAGA': 1, 'TGCCA': 1, 'AATCC': 1, 'GGCAG': 1, 'GCGGC': 1, 'CCGGA': 1, 'AGAGG': 1, 'AGAGA': 4, 'GTGCA': 1, 'TGGAA': 3, 'TCTCC': 1, 'TGGAT': 2, 'AGCGG': 2, 'GTGTG': 3, 'ATGAT': 1, 'CTGCT': 1, 'TTTAA': 1, 'ATATA': 1, 'CTTCC': 1, 'ATATT': 1, 'CCTGA': 1, 'TCTTT': 1, 'TTTAT': 2, 'GGTGG': 2, 'ATTTT': 4, 'TGTGT': 1, 'CTCCT': 1, 'AAAAT': 4, 'CTGTG': 1, 'GACAC': 1, 'AAAAA': 28, 'AAAAC': 1, 'ATCAG': 1, 'ATCAA': 1, 'CTCTC': 3, 'TGAGA': 1, 'TGCTG': 8, 'AAACA': 1, 'GGCGG': 2, 'ACACA': 6, 'TTTTT': 33, 'AGGAA': 2, 'TAAAA': 2, 'TTTTA': 1, 'GATGA': 1, 'TGGGT': 1, 'ACAAC': 1, 'TTTCT': 1, 'ATGGA': 2, 'CAGAA': 3, 'GGAGG': 2, 'ATAGA': 1, 'TATTT': 2}\n",
    "\n",
    "print len(temp)\n",
    "#print [len(temp[i]) for i in xrange(len(temp)) if len(temp[i]) != 100]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 0.59785714  0.42324877  0.04288089  0.90853147]\n",
      " [ 0.11213601  0.82429483  0.97159137  0.19462965]\n",
      " [ 0.32879053  0.86143494  0.81872     0.98790726]\n",
      " [ 0.18781606  0.84905971  0.1761741   0.5781587 ]]\n",
      "[ 1.  2.  3.  4.]\n",
      "Softmax.0\n",
      "2\n",
      "[[ 0.00319925  0.77707296  0.04252415  0.17720363]]\n",
      "[1]\n",
      "[[ 0.00319925  0.          0.04252415  0.17720363]]\n",
      "[3]\n"
     ]
    }
   ],
   "source": [
    "import numpy\n",
    "import theano\n",
    "import theano.tensor as T\n",
    "\n",
    "# x = T.ivector('input')\n",
    "# y = T.argmax(x)\n",
    "# get_argmax = theano.function([x], y)\n",
    "# arrTest = numpy.array([1,2,3,4,19,1])\n",
    "# print get_argmax(arrTest)\n",
    "\n",
    "# elem = tensor.scalar('elem')\n",
    "# elem = get_argmax(arrTest)\n",
    "# val = -np.inf\n",
    "# arrTest = tensor.set_subtensor(arrTest[elem:elem+1], val)\n",
    "\n",
    "# print arrTest\n",
    "\n",
    "W = theano.shared(\n",
    "            value=numpy.random.rand(4, 4),\n",
    "            name='W',\n",
    "            borrow=True\n",
    "        )\n",
    "\n",
    "input = theano.shared(numpy.asarray([1,2,3,4],\n",
    "                                dtype=theano.config.floatX),\n",
    "                                borrow=True)\n",
    "\n",
    "y_given_x = T.nnet.softmax(T.dot(input, W))\n",
    "\n",
    "y_pred = T.argmax(y_given_x, axis=1)\n",
    "\n",
    "print W.eval()\n",
    "print input.eval()\n",
    "print y_given_x\n",
    "print y_given_x.ndim\n",
    "print y_given_x.eval()\n",
    "print y_pred.eval()\n",
    "\n",
    "temp = y_given_x.eval()\n",
    "temp[0,y_pred.eval()] = 0\n",
    "y_given_x = theano.shared(temp)\n",
    "\n",
    "y_pred = T.argmax(y_given_x, axis=1)\n",
    "\n",
    "print y_given_x.eval()\n",
    "print y_pred.eval()\n",
    "\n",
    "\n",
    "# a = T.ivector('a')\n",
    "# b = T.ivector('b')\n",
    "# c = T.ivector('c')\n",
    "# out1 = T.or_(T.or_(T.eq(a,b),T.eq(a,c)),T.eq(b,c))\n",
    "#out2 = T.mean(T.eq(a,c))\n",
    "\n",
    "# get_out1 = theano.function([a,b,c], out1)\n",
    "#get_out2 = theano.function([a,c], out2)\n",
    "\n",
    "# ans1 = get_out1([1,2,3,4], [7,8,9,10], [1,2,9,6])\n",
    "#ans2 = get_out2([1,2,3,4], [1,2,5,6])\n",
    "\n",
    "# print ans1\n",
    "#print ans2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'theano.tensor.var.TensorVariable'>\n",
      "(1L, 4L)\n",
      "<type 'numpy.ndarray'>\n"
     ]
    }
   ],
   "source": [
    "print type(y_given_x)\n",
    "# print y_given_x.ndim\n",
    "print temp.shape\n",
    "print type(temp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
